script/transform-approved-books-data.ts

```diff
@@ -0,0 +1,178 @@
+  return typeof something.collection_id === 'string'
+    && typeof something.server === 'string'
+    && isApprovedBooksData(something);
+};
+
+interface ApprovedRepo extends ApprovedBooksData {
+  repo: string;
+}
+
+const isApprovedRepo = (something: any): something is ApprovedRepo => {
+  return typeof something.repo === 'string' && isApprovedBooksData(something);
+};
+
+interface ApprovedVersion {
+  content_version: string;
+  min_code_version: string;
+}
+
+const isApprovedVersion = (something: any): something is ApprovedVersion => {
+  return typeof something.content_version === 'string' && typeof something.min_code_version === 'string';
+};
+
+interface ApprovedVersionCollection extends ApprovedVersion {
+  collection_id: string;
+}
+
+const isApprovedVersionCollection = (something: any): something is ApprovedVersionCollection => {
+  return typeof something.collection_id === 'string' && isApprovedVersion(something);
+};
+
+interface ApprovedVersionRepo extends ApprovedVersion {
+  repo: string;
+}
+
+const isApprovedVersionRepo = (something: any): something is ApprovedVersionRepo => {
+  return typeof something.repo === 'string' && isApprovedVersion(something);
+};
+
+interface ApprovedBooksAndVersions {
+  approved_books: Array<ApprovedCollection | ApprovedRepo>;
+  approved_versions: Array<ApprovedVersionCollection | ApprovedVersionRepo>;
+}
+
+const isApprovedBooksAndVersions = (something: any): something is ApprovedBooksAndVersions => {
+  return Array.isArray(something.approved_books)
+    && something.approved_books.every(
+      (element: any) => isApprovedCollection(element) || isApprovedRepo(element))
+    && something.approved_versions.every(
+      (element: any) => isApprovedVersionCollection(element) || isApprovedVersionRepo(element));
+};
+
+const matchRepoVersion = (repoData: ApprovedRepo, archiveVersion?: string) =>
+  (versionData: ApprovedVersionCollection | ApprovedVersionRepo): versionData is ApprovedVersionRepo => {
+  if (archiveVersion && versionData.min_code_version !== archiveVersion) { return false; }
+  return isApprovedVersionRepo(versionData) && versionData.repo === repoData.repo;
+};
+
+const matchCollectionVersion = (collectionData: ApprovedCollection, archiveVersion?: string) =>
+  (versionData: ApprovedVersionCollection | ApprovedVersionRepo): versionData is ApprovedVersionCollection => {
+  if (archiveVersion && versionData.min_code_version !== archiveVersion) { return false; }
+  return isApprovedVersionCollection(versionData) && versionData.collection_id === collectionData.collection_id;
+};
+
+const getDesiredVersion = (
+  approvedVersions: Array<ApprovedVersionCollection | ApprovedVersionRepo>,
+  colOrRepo: ApprovedRepo | ApprovedCollection
+): string | undefined => {
+  // Example archive code version: 20210224.204120
+  const archiveVersionMatch = REACT_APP_ARCHIVE_URL.match(/[0-9]+\.[0-9]+/);
+  const archiveVersion = archiveVersionMatch ? archiveVersionMatch[0] : undefined;
```

Comment:
we should probably stop supporting this being undefined now. also, if it makes life easier we can have the archive version in a separate variable, but that would also require modifying the archive update script so if you want to just add an `assertDefined` here thats fine too